DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c

```diff
@@ -861,29 +865,392 @@ QueryVariableInfoHook (
+
+      // E. Blt to Screen at FIXED Left Margin
+      //    We now blit the entire "strip" which fits perfectly inside the box padding
+      Gop->Blt(Gop, RenderBufferScaled, EfiBltBufferToVideo,
+               0, 0,
+               TextStartX, CurrentY,          // Fixed X Position
+               UnscaledContentW * ScaleFactor, GLYPH_HEIGHT * ScaleFactor,
+               UnscaledContentW * ScaleFactor * sizeof(EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
+    }
+
+    CurrentY += (GLYPH_HEIGHT * ScaleFactor);
+  }
+  VA_END(ArgsCopy);
+
+  if (RenderBuffer1x) FreePool(RenderBuffer1x);
+  if (RenderBufferScaled) FreePool(RenderBufferScaled);
+  if (BltBufferObj)   FreePool(BltBufferObj);
+}
 
-  return !DescriptorWriteable;
+STATIC
+VOID
+HexDump (
+  IN UINT8 *InputData,
+  IN UINTN InputLen,
+  OUT CHAR16 *OutputStr
+  )
+{
+  STATIC CONST CHAR8 HexMap[] = "0123456789ABCDEF";
+  for (UINTN i = 0; i < InputLen; i++) {
+    OutputStr[i * 2]     = HexMap[InputData[i] >> 4];
+    OutputStr[i * 2 + 1] = HexMap[InputData[i] & 0x0F];
+  }
+  OutputStr[InputLen * 2] = '\0';
+}
+
+STATIC
+UINTN
+GetDisplayScaleFactor (
+  VOID
+  )
+{
+EFI_STATUS                             Status;
+  EFI_GRAPHICS_OUTPUT_PROTOCOL         *Gop;
+  UINT32                               ModeIndex;
+  UINT32                               MaxHRes = 0;
+  EFI_GRAPHICS_OUTPUT_MODE_INFORMATION *Info;
+  UINTN                                SizeOfInfo;
+
+  Status = gBS->LocateProtocol (&gEfiGraphicsOutputProtocolGuid, NULL, (VOID **)&Gop);
+  if (EFI_ERROR (Status) || Gop == NULL)
+    return 1;
+
+  // Iterate through all modes to find the largest (presumed native) resolution
+  for (ModeIndex = 0; ModeIndex < Gop->Mode->MaxMode; ModeIndex++) {
+    Status = Gop->QueryMode (Gop, ModeIndex, &SizeOfInfo, &Info);
+
+    if (!EFI_ERROR (Status)) {
+      if (Info->HorizontalResolution > MaxHRes) {
+        MaxHRes = Info->HorizontalResolution;
+      }
+      FreePool (Info);
+    }
+  }
+
+  if (MaxHRes > 1920)
+    return 2;
+
+  return 1;
+}
+
+STATIC
+VOID
+ShowBtgErrorPopup (
+  IN VOID   *CurrentImage,
+  IN UINTN  ImageSize,
+  IN CHAR16 *ErrorMessage
+  )
+{
+  EFI_STATUS     Status;
+  UINT8          *CurrentOemRk;
+  UINT16         CurrentOemRkString[48 * 2 + 1];
+  UINTN          CurrentAttribute;
+  BOOLEAN        CursorVisible;
+  EFI_EVENT      Events[1];
+  UINTN          Index;
+  EFI_INPUT_KEY  Key;
+
+  Status = GetOemRootKeyHash(CurrentImage, ImageSize, &CurrentOemRk);
+  ASSERT_EFI_ERROR (Status);
+
+  HexDump (CurrentOemRk, 48, CurrentOemRkString);
+
+  CurrentAttribute = gST->ConOut->Mode->Attribute;
+  CursorVisible    = gST->ConOut->Mode->CursorVisible;
+
+  gST->ConOut->EnableCursor (gST->ConOut, FALSE);
+
+  DrainInput ();
+
+  Events[0] = gST->ConIn->WaitForKey;
+
+  do {
+    ClearScreen();
+    CreateMultiStringPopUpScaled (
+        GetDisplayScaleFactor(),
+        100,

```

Comment:
```suggestion
        96,
```
With `100`:
```
ScreenWidth=800
ScreenHeight=600
RequestedWidth=100
UnscaledContentW=800
UnscaledContentH=247
BoxW=832
BoxH=285
```

<img width="1600" height="1248" alt="Image" src="https://github.com/user-attachments/assets/a15d9add-b23f-4dae-a886-3b66848ec82b" />

While `96` fits perfectly:
```
RequestedWidth=96
ScreenWidth=800
ScreenHeight=600
RequestedWidth=96
UnscaledContentW=768
UnscaledContentH=247
BoxW=800
BoxH=285
```

<img width="1600" height="1248" alt="Image" src="https://github.com/user-attachments/assets/cf0feae1-9c2f-4fc1-a27a-1625a092967c" />

Reply:
https://github.com/Dasharo/edk2/pull/285/commits/b3f0caeafcb3cde4d3bf8d39d1cfae12a7c9fb10